Strength 2 Covering Array constructions

aadwyer opened the pull request:
This patch aims to add functionality to the covering arrays module that was recently added to SageMath. It adds methods that create many covering arrays of strength 2 that can later be used for recursive constructions as well.

Specifically: added methods to generate covering arrays and a main CA function to call them based on inputted parameters. 
Added methods include direct implementation of optimal arrays in the database of small combinatorial designs, methods that create a CA(N,t=2,k,v=2), for either k or N inputted and a recursive method that removes columns from a larger array. Added required documentation and reference for new methods.

This fixes #37371

aadwyer committed:
Added methods to generate covering arrays and a main CA function to
call them based on inputted paramters. Added methods include direct
implementation of optimal arrays in the database of small
combinatorial designs, methods that create a CA(N,t=2,k,v=2), for
either k or N inputted and a recursive method that removes columns
from a larger array.

Added required documentation and reference for new methods.

aadwyer committed:
Fixed formatting issues

aadwyer committed:
Formatting issues part 2

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+        (True, (5, 2, 4, 2))
+
+    """
+    from math import ceil
```
Consider using `sage.arith.misc.integer_ceil`

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+
+    - ``strength`` (integer) -- the parameter `t` of the covering array,
+      such that in any selection of `t` columns of the array, every `t`
+      -tuple appears at least once.
```
```suggestion
      such that in any selection of `t` columns of the array, every 
      `t`-tuple appears at least once.
```

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+def column_Kleitman_Spencer_Katona(k):
+    r"""
+    Return a covering array with `k` columns using the Kleitman Spencer Katona
+    method.
```
```suggestion
    Return a covering array with `k` columns using the Kleitman-Spencer-Katona
    method.
```

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+                CA = "ca_{}_{}_{}_{}".format(i[0], strength, i[1], levels)
+                f = getattr(DB, CA)
+                return truncate_columns(f(), number_columns)
+
```
can this case fall through here?

sraaphorst replied:
When you say "fall through," what do you mean? I'm trying to understand the comment.

aadwyer replied:
I am not sure if I understand either, hopefully the changes I made in the recent patch resolve this. If not please let me know what you mean.

mkoeppe replied:
"Fall through" == a case in which the body of the `if` is not terminated by `return`.

sraaphorst replied:
Ah. I see what you mean now. Yes, if the `for` terminates and no `return` is triggered, the code doesn't reach a `return` at all.

aadwyer replied:
Thank you, there definitely was a problem with the code at that point but the current version does not fall through when 3<v<6.

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+    if strength != 2:
+        raise ValueError("CAs and OAs only implemented for strength 2")
+
+    elif levels == 2:
```
```suggestion
    if levels == 2:
```
because the `if` branch always raises

sraaphorst replied:
Right. No point in having an `elif` here.

I'm aslo not sure why `assert` is being used in the preceding function, but exceptions are being used in this function. One or the other should be picked.

I would probably change the assert in the previous function to an `if ...: raise ValueError(...)`.

mkoeppe commented on src/sage/combinat/designs/database.py:
```diff
@@ -5033,6 +5035,462 @@ def BIBD_56_11_2():
+
+def ca_11_2_5_3():
+    r"""
+    Return a CA with the given parameters, This CA is proven to be optimal.
```
```suggestion
    Return a CA with the given parameters. This CA is proven to be optimal.
```

(same several times below)

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+
+def Kleitman_Spencer_Katona(N):
+    r"""
+    Return a `CA(N; 2, k, 2)` where `k = \binom {N-1}{\lceil \frac{N}{2} \rceil}`.
```
```suggestion
    Return a `CA(N; 2, k, 2)` where `k = \binom {N-1}{\lceil N/2 \rceil}`.
```
maybe better to use inline division?

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+    assert k <= 24310, "not implemented for k > 24310"
+
+    kdict = {3:4,4:5,10:6,15:7,35:8,56:9,126:10,210:11,462:12,792:13,
+            1716:14,3003:15,6435:16,11440:17,24310:18}
```
```suggestion
    kdict = {3: 4, 4: 5, 10: 6, 15: 7, 35: 8, 56: 9, 126: 10, 210: 11, 462: 12, 792: 13,
            1716: 14, 3003: 15, 6435: 16, 11440: 17, 24310: 18}
```

sraaphorst commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+        for row in array:
+            result.append(row[:k])
+        return result
+
```
I'm not sure about the Sage guideline, but in the interests of efficiency, the `else` part is probably better done:
```
return [row[:k] for row in array]
```
Then the looping happens behind the hood in C instead of in Python. For objects of these sizes, it probably doesn't matter, though.
for efficiency. It's best to

sraaphorst commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+        for i in p:
+            S[i] = 1
+        result.append(S)
+    return(list(map(list, zip(*result))))
```
The logic here is correct, but the last line is confusing. You might want to build the actual matrix instead of its transpose, or at least add a line before the last line saying that you are taking the transpose. `map` is generally not used much in Python 3+ since it is typically considerably slower than just using list comprehensions.

sraaphorst replied:
Also, I don't know about SageMath conventions, but in Python, the convention is to not use caps in function names. Just thought I'd point it out.

aadwyer replied:
I am not sure, the style guide does say the same thing, but other functions in sage capitalize names I believe.

sraaphorst commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+        if k <= ki:
+            N = kdict[ki]
+            break
+    return truncate_columns(Kleitman_Spencer_Katona(N), k)
```
The `kdict` approach may not work: Python does not guarantee that dictionaries maintain their order. If it places the 24310 at the beginning, for example, you will always get 18.

I would implement this as a list of tuple:
```
kdict = [(3, 4), (4, 5), (10, 6), (15, 7), (35, 8), (56, 9), (126, 10), (210, 11), (462, 12), (792, 13), (1716, 14), (3003, 15), (6435, 16), (11440, 17), (24310, 18)]
```

I also wouldn't hard-copy the maximum value of `k` in the code like you do above. If you can ever extend `kdict`, then you'll have to change it in two places. I would do (after the above definition):

```
assert k <= kdict[-1][0]
```

Then the rest of the logic turns into:

```
for ki, N in kdict:
    if k <= ki:
        return truncate_columns(Kleitman_Spencer_Katona(N), k)
```

Depending on the version of Python you're using, all of this can probably be replaced by the `bisect` library, using, I believe, the `bisect_left` method.

aadwyer replied:
The first suggestion is much appreciated. Also bisect_left was very interesting to learn about but I am not sure what is faster when the lists are so small like this, even if it is extended I do not imagine it will be extended very much.

mkoeppe replied:
> Python does not guarantee that dictionaries maintain their order.

Not true any more since Python 3.7. Dictionary maintain insertion order

sraaphorst replied:
Yes, you are correct, @aadwyer: maps are now ordered.

aadwyer replied:
Because there are only a small number of entries, the NotImplementedError is raised fast and it will often have to read each entry in either case, I will keep this as a list despite the fact dictionaries could work now.

sraaphorst reviewed (changes requested):
This looks good in terms of logic. Some changes should be made for standardization and correctness.

aadwyer committed:
Added an in_database function, it now allows to check the database
more generally so it does not need to be changed if things are
added. Also edited some functions to produce the same output
more efficiently. Fixed some formatting and punctuation errors.

aadwyer committed:
Merge branch 'sagemath:develop' into covering_arrays

aadwyer committed:
Merge branch 'develop' into strength2_methods

aadwyer committed:
Merge branch 'strength2_methods' into covering_arrays

aadwyer committed:
Merge pull request #21 from aadwyer/covering_arrays

Covering arrays

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -50,3 +54,190 @@
+        AssertionError: not implemented for k > 24310
+
+    """
+    assert k <= 24310, "not implemented for k > 24310"
```
This should be an `if... raise NotImplementedError(...)`

mkoeppe commented on src/sage/combinat/designs/database.py:
```diff
@@ -5033,6 +5035,462 @@ def BIBD_56_11_2():
+    (2,4) : ((19,6), (21,7)),
+    (2,5) : ((29,7),),
+    (2,6) : ((37,4), (39,5), (41,6))
+}
```
```suggestion
CA_constructions = {
    (2,3): ((11,5), (12,7), (13,9), (14,10), (15,20)),
    (2,4): ((19,6), (21,7)),
    (2,5): ((29,7),),
    (2,6): ((37,4), (39,5), (41,6))
}
```
coding style

aadwyer committed:
Update covering_array.py

Changed ValueError to NotImplementedError

aadwyer committed:
Update src/sage/combinat/designs/database.py

Co-authored-by: Matthias Köppe <mkoeppe@math.ucdavis.edu>

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -49,3 +54,216 @@
+    if oldk == k:
+        return array
+
+    elif oldk < k:
```
In an `if ... return`, best to continue with just `if`, not `elif`

sraaphorst replied:
I agree with this comment. This should probably be written as:

```python
oldk = len(array[0])

if oldk == k:
    return array

if oldk < k:
    raise ValueError("...")

return [row[:k] for row in array]
```

Handle the exceptional cases first that return trivial results and handle bad input, and then present the default case.

mkoeppe commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -49,3 +54,216 @@
+    if in_database:
+        return in_database
+
+    if orthogonal_array(number_columns, levels, strength, existence=True) == True:
```
see message from Ruff

aadwyer replied:
https://github.com/sagemath/sage/pull/37372#issuecomment-2174348903

I noticed this but I am not sure how to proceed to make it happy while also keeping the functionality we need.

mkoeppe replied:
```suggestion
    if orthogonal_array(number_columns, levels, strength, existence=True) is True:
```

aadwyer committed:
Update src/sage/combinat/designs/covering_array.py

Co-authored-by: Matthias Köppe <mkoeppe@math.ucdavis.edu>

aadwyer committed:
Changed the elifs in truncate columns

mkoeppe reviewed (approved):
LGTM
